core/trino-spi/src/main/java/io/trino/spi/predicate/Range.java

```diff
@@ -40,165 +69,280 @@ public Range(Marker low, Marker high)
+            throw new IllegalArgumentException("low bound must be exclusive for low unbounded range");
+        }
+        if (highValue.isEmpty() && highInclusive) {
+            throw new IllegalArgumentException("high bound must be exclusive for high unbounded range");
+        }
+        boolean isSingleValue = false;
+        if (lowValue.isPresent() && highValue.isPresent()) {
+            int compare = compareValues(comparisonOperator, lowValue.get(), highValue.get());
+            if (compare > 0) {
+                throw new IllegalArgumentException("low must be less than or equal to high");
+            }
+            if (compare == 0) {
+                if (!highInclusive || !lowInclusive) {
+                    throw new IllegalArgumentException("invalid bounds for single value range");
+                }
+                isSingleValue = true;
+            }
+        }
+        lowValue.ifPresent(value -> verifyNotNan(type, value));
+        highValue.ifPresent(value -> verifyNotNan(type, value));
+
+        this.lowInclusive = lowInclusive;
+        this.lowValue = lowValue;
+        this.highInclusive = highInclusive;
+        this.highValue = highValue;
+
+        this.comparisonOperator = comparisonOperator;
+        this.isSingleValue = isSingleValue;
+    }
+
+    private static void verifyNotNan(Type type, Object value)
+    {
+        if (isFloatingPointNaN(type, value)) {
+            throw new IllegalArgumentException("cannot use NaN as range bound");
         }
-        this.low = low;
-        this.high = high;
     }
 
     public static Range all(Type type)
     {
-        return new Range(Marker.lowerUnbounded(type), Marker.upperUnbounded(type));
+        return new Range(type, false, Optional.empty(), false, Optional.empty());
     }
 
     public static Range greaterThan(Type type, Object low)
     {
-        return new Range(Marker.above(type, low), Marker.upperUnbounded(type));
+        return new Range(type, false, Optional.of(low), false, Optional.empty());
     }
 
     public static Range greaterThanOrEqual(Type type, Object low)
     {
-        return new Range(Marker.exactly(type, low), Marker.upperUnbounded(type));
+        return new Range(type, true, Optional.of(low), false, Optional.empty());
     }
 
     public static Range lessThan(Type type, Object high)
     {
-        return new Range(Marker.lowerUnbounded(type), Marker.below(type, high));
+        return new Range(type, false, Optional.empty(), false, Optional.of(high));
     }
 
     public static Range lessThanOrEqual(Type type, Object high)
     {
-        return new Range(Marker.lowerUnbounded(type), Marker.exactly(type, high));
+        return new Range(type, false, Optional.empty(), true, Optional.of(high));
     }
 
     public static Range equal(Type type, Object value)
     {
-        Marker marker = Marker.exactly(type, value);
-        return new Range(marker, marker);
+        requireNonNull(value, "value is null");
```

Comment:
why check here but not in factory method above

Reply:
Added rnn checks to all